lib/partition_setup.pm

```diff
@@ -438,37 +438,29 @@ sub addboot {
     }
 }
 
-=head2 skip_select_first_hard_disk
-
- skip_select_first_hard_disk();
-
-Skip selecting first hard disk and return 0.
-Return 1 if no selection of hard disk is required or the first hard disk is already pre-selected.
-
-=cut
-sub skip_select_first_hard_disk {
-    return 1 if match_has_tag 'existing-partitions';          # no selection of hard-disk is required
-    if (match_has_tag('select-hard-disks-one-selected')) {    # first hd is already pre-selected
-        send_key $cmd{next};
-        return 1;
-    }
-    return 0;
-}
-
 =head2 select_first_hard_disk
 
  select_first_hard_disk();
 
 Select the first hard disk.
 
+The device should be [sv]da, other devices will be unselected. [sv]da device will also be
+force-selected at the end if needed (in some cases [sv]da is at the end of the list).
+
 =cut
 sub select_first_hard_disk {
-    my @tags = qw(existing-partitions hard-disk-dev-sdb-selected hard-disk-dev-sdc-selected hard-disk-dev-sdd-selected hard-disk-dev-non-sda-selected select-hard-disks-one-selected);
-    my $matched_needle = assert_screen \@tags;
-    return if skip_select_first_hard_disk;
+    # Try to handle most of the device type
+    my @tags    = 'existing-partitions';
+    my @devices = ('sdb' .. 'sdz', 'vdb' .. 'vdz', 'pmem0' .. 'pmem9');
+    foreach my $device (@devices) {
+        push @tags, "hard-disk-dev-$device-selected";
+    }
+    my $matched_needle = check_screen \@tags;           # save detected needle
+    return 1 if match_has_tag 'existing-partitions';    # no selection of hard-disk is required
```

Comment:
Previous code did not have an early exit when matching `existing-partitions`, only on `select-hard-disks-one-selected`. Wouldn't this early exit here avoid re-partitioning/re-formatting the first disk on installations over IPMI where an OS was successfully installed previously?

Reply:
You're right, I will check again and correct this.

Reply:
At the end maybe I missed something, but I did a test and in the previous code if `existing-partitions` is matched a `return 1` is executed in function `skip_select_first_hard_disk`. In function `select_first_hard_disk` we have this code `return if skip_select_first_hard_disk;`, so the function ends if `skip_select_first_hard_disk` return true, no? In that case if `select-hard-disks-one-selected` is found as you said, but also if `existing-partitions` is found, no?

Reply:
Oh, you're right. My bad